### PR TITLE
REF:added refactoring to convert tuple to struct

### LINKED ISSUE #3180: Add "convert to block/typle struct" intention
Add 2 intentions to convert  structs between block/tuple fields:

```rust
struct Foo(i32, i32);
```
<=>
```rust
struct Foo {
    _0: i32,
    _1: i32,
}
```

### PR DESCRIPTION
**Pull request by rrevenantt**
Closes #3180 

I have also refactored code a bit.

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/refactoring/convertStruct/RsConvertToTupleProcessor.kt [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import org.rust.lang.core.psi.ext.*
+
+class RsConvertToTupleProcessor(project: Project, val element: RsFieldsOwner, val convertUsages: Boolean = true) : BaseRefactoringProcessor(project) {
+    private val rsPsiFactory = RsPsiFactory(project)
+    private val fieldDeclList = element.blockFields!!.namedFieldDeclList
+    override fun findUsages(): Array<UsageInfo> {
+        if (!convertUsages) return arrayOf()
+
+        var usages = ReferencesSearch
+            .search(element, element.useScope)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/refactoring/convertStruct/RsConvertToTupleProcessor.kt]: As far as I know, `element.useScope` isn't reasonable here b/c `ReferencesSearch` internally retrieves `element.useScope` end intersects in with an argument scope

File: src/main/kotlin/org/rust/ide/refactoring/convertStruct/RsConvertToTupleProcessor.kt [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            .asSequence()
+            .map { UsageInfo(it) }
+
+        usages += fieldDeclList
+            .mapIndexed { index, rsNamedFieldDecl ->
+                ProgressManager.checkCanceled()
+                ReferencesSearch
+                    .search(rsNamedFieldDecl, rsNamedFieldDecl.useScope)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/refactoring/convertStruct/RsConvertToTupleProcessor.kt]: ... and here

File: src/main/kotlin/org/rust/ide/refactoring/convertStruct/RsConvertToTupleProcessor.kt [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                    //other references will be handled from main struct usages
+                    .filter { it.element.parent is RsDotExpr }
+                    .map { MyUsageInfo(it, index) }
+            }.flatten()
+
+        return usages
+            .toCollection(ArrayList())
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/refactoring/convertStruct/RsConvertToTupleProcessor.kt]: Is it necessary?

File: src/test/kotlin/org/rust/ide/refactoring/ConvertToTupleRefactoringTest.kt [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        fn main (){
+            let Test(a, b) = Test(0, 0);
+            let Test(ref a, mut var) = Test(0, 0);
+            let Test(var1, var2) = Test(0, 0);
+            let Test(a, _) = Test(0, 0);
+            let Test(_, var) = Test(0, 0);
+        }
+    """)
+
+    protected fun doAvailableTest(@Language("Rust") before: String, @Language("Rust") after: String) {
```
Comment (vlad20012) [inline src/test/kotlin/org/rust/ide/refactoring/ConvertToTupleRefactoringTest.kt]: Let's add a corner case with empty field list: `struct S {}`.

File: src/main/kotlin/org/rust/ide/refactoring/convertStruct/RsConvertToTupleProcessor.kt (hunk 3 above)
Comment (rrevenantt) [reply src/main/kotlin/org/rust/ide/refactoring/convertStruct/RsConvertToTupleProcessor.kt]: Sequence doesn't have `toArray` or similar, does it?

Comment (vlad20012) [reply src/main/kotlin/org/rust/ide/refactoring/convertStruct/RsConvertToTupleProcessor.kt]: Umm `.toList()`? `.toCollection(ArrayList())` looks very weird in kotlin

Comment (rrevenantt) [reply src/main/kotlin/org/rust/ide/refactoring/convertStruct/RsConvertToTupleProcessor.kt]: Oh, indeed.

**Comment by vlad20012** [comment]
Thanks!
bors r+
